### PR TITLE
ILM Prefactor

### PR DESCRIPTION
**Pull request by sravfeyn**
In preperation for ILM based ES indices on ICDS

@esoergel

### CODE REVIEW

File: corehq/ex-submodules/pillowtop/es_utils.py [hunk 1]
```diff
@@ -136,22 +137,22 @@ def set_index_normal_settings(es, index):
...
-    initialize_mapping_if_necessary(es, index_info)
     assume_alias(es, index_info.index, index_info.alias)
 
 
 def initialize_index(es, index_info):
-    return create_index_and_set_settings_normal(es, index_info.index, index_info.meta)
+    index = index_info.index
+    mapping = transform_for_es7(index_info.mapping)
+    mapping['_meta']['created'] = datetime.isoformat(datetime.utcnow())
+    meta = index_info.meta
```
Comment (sravfeyn) [inline corehq/ex-submodules/pillowtop/es_utils.py:151]: Todo; make a copy

**Comment by dannyroberts** [comment]
FYI travis tests not passing

**Comment by sravfeyn** [comment]
@esoergel @snopoke 

This is now ready for review and merge. This is refactor portion of https://github.com/dimagi/commcare-hq/pull/28574

**Comment by sravfeyn** [comment]
Bumping for review 

(cc @millerdev code-buddy)

File: corehq/ex-submodules/pillowtop/tests/test_elasticsearch.py [hunk 2]
```diff
@@ -91,7 +91,8 @@ def test_assume_alias(self):
         initialize_index_and_mapping(self.es, TEST_INDEX_INFO)
         doc_id = uuid.uuid4().hex
         doc = {'_id': doc_id, 'doc_type': 'CommCareCase', 'type': 'mother'}
-        send_to_elasticsearch(self.index, TEST_INDEX_INFO.type, doc_id, get_es_new, 'test', doc)
+        ElasticsearchInterface(get_es_new()).create_doc(
+            self.index, TEST_INDEX_INFO.type, doc_id, {'doc_type': 'CommCareCase', 'type': 'mother'}, False)
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/tests/test_elasticsearch.py:95]: why this change?

**Review by snopoke — COMMENTED** [review]
One question otherwise changes look fine.

File: corehq/ex-submodules/pillowtop/tests/test_elasticsearch.py (hunk 2 above)
Comment (sravfeyn) [reply corehq/ex-submodules/pillowtop/tests/test_elasticsearch.py:95]: I wanted to make `pillowtop.processors.elastic.send_to_elasticsearch` more private (preferring the one in `corehq.elastic.send_to_elasticsearch` instead. We could use the one `corehq.elastic.send_to_elasticsearch` here, but it requires mocking `ES_MET`, so I made this change.

Comment (sravfeyn) [reply corehq/ex-submodules/pillowtop/tests/test_elasticsearch.py:95]: This should be fine, since `ElasticsearchInterface` is designed for public usage.

File: .travis.yml [hunk 3]
```diff
@@ -37,7 +37,8 @@ script:
   - export TRAVIS_HQ_USERNAME=$TRAVIS_HQ_USERNAME
   - export TRAVIS_HQ_PASSWORD=$TRAVIS_HQ_PASSWORD
   - export DATADOG_API_KEY=$DATADOG_API_KEY
-  - scripts/docker test --noinput --stop --verbosity=2 --divide-depth=1 --with-timing --threshold=10 --max-test-time=29
+  # todo add --stop
+  - scripts/docker test --noinput --verbosity=2 --divide-depth=1 --with-timing --threshold=10 --max-test-time=29
```
Comment (millerdev) [inline .travis.yml:41]: This change should not be merged

**Review by millerdev — CHANGES_REQUESTED** [review]

**Review by millerdev — APPROVED** [review]

**Comment by sravfeyn** [comment]
@snopoke This is good to go?

File: corehq/ex-submodules/pillowtop/processors/elastic.py [hunk 4]
```diff
@@ -159,25 +161,28 @@ def process_changes_chunk(self, changes_chunk):
...
+    retries = 1 if settings.UNIT_TESTING else MAX_RETRIES
+    propagate_failure = settings.UNIT_TESTING
     while current_tries < retries:
         try:
+            doc_exists = es_interface.doc_exists(alias, doc_id, doc_type)
             if delete:
-                es_interface.delete_doc(alias, doc_type, doc_id)
-            elif update:
+                if doc_exists:
+                    es_interface.delete_doc(alias, doc_type, doc_id)
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/processors/elastic.py:184]: thought for future: what about just doing the delete here without the check to see if it exists and handling the 404 error?

File: corehq/ex-submodules/pillowtop/processors/elastic.py [hunk 5]
```diff
@@ -159,25 +161,28 @@ def process_changes_chunk(self, changes_chunk):
...
+            elif doc_exists:
                 params = {'retry_on_conflict': 2}
                 if es_merge_update:
                     es_interface.update_doc_fields(alias, doc_type, doc_id, fields=data, params=params)
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/processors/elastic.py:188]: another thought for future:

The only difference between `create` and `index` is that create will raise an error if the doc exists. I don't think we depend on that anywhere so we could just use the `index` method for both create and update (unless we need to 'update_fields'). If we did that (and my other thought above) we wouldn't need to check for doc existence at all except in cases where we want to only do a `fields` update.

Furthermore, in the case of `es_merge_update` we are assuming that the doc already exists so we could just call `update` without the check. If the doc doesn't exist we get a 404 which we can handle. The current functionality will create a new doc with only the fields specified which I think is incorrect behaviour.

**Review by snopoke — APPROVED** [review]
Current changes look fine. I've made some comments for potential future changes.

**Comment by sravfeyn** [comment]
Thanks for the comments Simon, since this is just prefactor, I will consider these in the actual ILM PR.
